[MIPS] Add FeatureMSA to i6400 and i6500 cores

mgoudar opened the pull request:
- Enable 'FeatureMSA' for MIPS i6400 and i6500 cpu.
- Enable -mmsa option if mcpu is set to either i6400 or i6500
- added clang driver test to validate msa feature
- added llvm codegen test to validate msa instructions for cpu i6500 and i6400

MIPS i6400 and i6500 cores implements and enables MSA (MIPS SIMD ARCHITECTURE) by default.

mgoudar committed:
[MIPS] Add FeatureMSA to i6400 and i6500 cores

i6400 and i6500 cores support MIPS SIMD Architecture (MSA) instructions

mshockwave reviewed (commented):
CI has reported an error:
```
Exit Code: 2

Command Output (stderr):
--
/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/llc -mtriple=mips-elf -mcpu=i6500 < /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/test/CodeGen/Mips/msa/i6500.ll |    /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/FileCheck /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/test/CodeGen/Mips/msa/i6500.ll --check-prefix=MIPS32 # RUN: at line 3
+ /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/llc -mtriple=mips-elf -mcpu=i6500
+ /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/FileCheck /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/test/CodeGen/Mips/msa/i6500.ll --check-prefix=MIPS32
Cannot handle this ValVT.
UNREACHABLE executed at /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/MipsISelLowering.cpp:3119!
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/llc -mtriple=mips-elf -mcpu=i6500
1.	Running pass 'Function Pass Manager' on module '<stdin>'.
2.	Running pass 'MIPS DAG->DAG Pattern Instruction Selection' on function '@llvm_mips_dlsa_test'
 #0 0x00005d08c95dee68 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Support/Unix/Signals.inc:804:13
 #1 0x00005d08c95dc99e llvm::sys::RunSignalHandlers() /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Support/Signals.cpp:106:18
 #2 0x00005d08c95df4f1 SignalHandler(int, siginfo_t*, void*) (/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/llc+0x74b54f1)
 #3 0x00007e830e867520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
 #4 0x00007e830e8bb9fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
 #5 0x00007e830e867476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
 #6 0x00007e830e84d7f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
 #7 0x00005d08c9549910 llvm::install_out_of_memory_new_handler() /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Support/ErrorHandling.cpp:195:0
 #8 0x00005d08c74ef031 CC_MipsO32(unsigned int, llvm::MVT, llvm::MVT, llvm::CCValAssign::LocInfo, llvm::ISD::ArgFlagsTy, llvm::CCState&, llvm::ArrayRef<unsigned short>) /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/MipsISelLowering.cpp:3099:7
 #9 0x00005d08c74e3284 CC_MipsO32_FP /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/lib/Target/Mips/MipsGenCallingConv.inc:381:9
#10 0x00005d08c74e3284 CC_Mips_FixedArg(unsigned int, llvm::MVT, llvm::MVT, llvm::CCValAssign::LocInfo, llvm::ISD::ArgFlagsTy, llvm::CCState&) /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/lib/Target/Mips/MipsGenCallingConv.inc:581:10
#11 0x00005d08c8496026 llvm::CCState::AnalyzeFormalArguments(llvm::SmallVectorImpl<llvm::ISD::InputArg> const&, bool (*)(unsigned int, llvm::MVT, llvm::MVT, llvm::CCValAssign::LocInfo, llvm::ISD::ArgFlagsTy, llvm::CCState&)) /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/CodeGen/CallingConvLower.cpp:92:9
#12 0x00005d08c74e9d80 getStackSize /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/include/llvm/CodeGen/CallingConvLower.h:245:42
#13 0x00005d08c74e9d80 llvm::MipsTargetLowering::LowerFormalArguments(llvm::SDValue, unsigned int, bool, llvm::SmallVectorImpl<llvm::ISD::InputArg> const&, llvm::SDLoc const&, llvm::SelectionDAG&, llvm::SmallVectorImpl<llvm::SDValue>&) const /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/MipsISelLowering.cpp:3798:35
```

mshockwave commented on llvm/test/CodeGen/Mips/msa/i6500.ll:
```diff
@@ -0,0 +1,69 @@
+; Test the MSA intrinsics that are encoded with the SPECIAL instruction format.
```
Could we use `update_llc_test_checks.py` to maintain this test?

mgoudar replied:
I have written driver test mips-cpus.c instead of codegen as you suggested.

mgoudar replied:
I have enabled MSA feature from driver also when -mcpu i6500/i6400 is specified. Also I have added codegen test to verify MSA instructions when i650/i6400 cpu is specified. please review

mshockwave reviewed (commented):
just a non-blocking suggestion: it'll be nice if you test individual CPU's features like https://github.com/llvm/llvm-project/blob/main/clang/test/Driver/riscv-cpus.c

mgoudar committed:
Update test case with update_llc_test_checks.py

mgoudar commented:
> CI has reported an error:
> 
> ```
> Exit Code: 2
> 
> Command Output (stderr):
> --
> /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/llc -mtriple=mips-elf -mcpu=i6500 < /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/test/CodeGen/Mips/msa/i6500.ll |    /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/FileCheck /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/test/CodeGen/Mips/msa/i6500.ll --check-prefix=MIPS32 # RUN: at line 3
> + /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/llc -mtriple=mips-elf -mcpu=i6500
> + /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/FileCheck /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/test/CodeGen/Mips/msa/i6500.ll --check-prefix=MIPS32
> Cannot handle this ValVT.
> UNREACHABLE executed at /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/MipsISelLowering.cpp:3119!
> PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
> Stack dump:
> 0.	Program arguments: /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/llc -mtriple=mips-elf -mcpu=i6500
> 1.	Running pass 'Function Pass Manager' on module '<stdin>'.
> 2.	Running pass 'MIPS DAG->DAG Pattern Instruction Selection' on function '@llvm_mips_dlsa_test'
>  #0 0x00005d08c95dee68 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Support/Unix/Signals.inc:804:13
>  #1 0x00005d08c95dc99e llvm::sys::RunSignalHandlers() /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Support/Signals.cpp:106:18
>  #2 0x00005d08c95df4f1 SignalHandler(int, siginfo_t*, void*) (/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/bin/llc+0x74b54f1)
>  #3 0x00007e830e867520 (/lib/x86_64-linux-gnu/libc.so.6+0x42520)
>  #4 0x00007e830e8bb9fc pthread_kill (/lib/x86_64-linux-gnu/libc.so.6+0x969fc)
>  #5 0x00007e830e867476 gsignal (/lib/x86_64-linux-gnu/libc.so.6+0x42476)
>  #6 0x00007e830e84d7f3 abort (/lib/x86_64-linux-gnu/libc.so.6+0x287f3)
>  #7 0x00005d08c9549910 llvm::install_out_of_memory_new_handler() /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Support/ErrorHandling.cpp:195:0
>  #8 0x00005d08c74ef031 CC_MipsO32(unsigned int, llvm::MVT, llvm::MVT, llvm::CCValAssign::LocInfo, llvm::ISD::ArgFlagsTy, llvm::CCState&, llvm::ArrayRef<unsigned short>) /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/MipsISelLowering.cpp:3099:7
>  #9 0x00005d08c74e3284 CC_MipsO32_FP /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/lib/Target/Mips/MipsGenCallingConv.inc:381:9
> #10 0x00005d08c74e3284 CC_Mips_FixedArg(unsigned int, llvm::MVT, llvm::MVT, llvm::CCValAssign::LocInfo, llvm::ISD::ArgFlagsTy, llvm::CCState&) /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/build/lib/Target/Mips/MipsGenCallingConv.inc:581:10
> #11 0x00005d08c8496026 llvm::CCState::AnalyzeFormalArguments(llvm::SmallVectorImpl<llvm::ISD::InputArg> const&, bool (*)(unsigned int, llvm::MVT, llvm::MVT, llvm::CCValAssign::LocInfo, llvm::ISD::ArgFlagsTy, llvm::CCState&)) /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/CodeGen/CallingConvLower.cpp:92:9
> #12 0x00005d08c74e9d80 getStackSize /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/include/llvm/CodeGen/CallingConvLower.h:245:42
> #13 0x00005d08c74e9d80 llvm::MipsTargetLowering::LowerFormalArguments(llvm::SDValue, unsigned int, bool, llvm::SmallVectorImpl<llvm::ISD::InputArg> const&, llvm::SDLoc const&, llvm::SelectionDAG&, llvm::SmallVectorImpl<llvm::SDValue>&) const /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-h94vh-1/llvm-project/github-pull-requests/llvm/lib/Target/Mips/MipsISelLowering.cpp:3798:35
> ```

yes. let me check it

mgoudar committed:
Add test case to test MSA feature for i6400 and i6500 MIPS cpus

Also remove codegen test i6500.ll

mshockwave commented on clang/test/Driver/mips-cpus.c:
```diff
@@ -0,0 +1,9 @@
+// Check target CPUs are correctly passed.
+
+// RUN: %clang --target=mips64 -### -c %s 2>&1 -mcpu=i6400 -mmsa | FileCheck -check-prefix=MCPU-I6400 %s
+// MCPU-I6400: "-target-cpu" "i6400"
```
usually we will check all the features implied by this CPU. Not sure if this is practical in your case though.

mgoudar replied:
I modified test to verify abicalls feature as well. I see that mips-features.c test available to verify other features based on the options

mgoudar committed:
Enable MSA feature by default for MIPS cpus i6400 and i6500

MIPS cpus support MSA by default. So enable this feature
for cpus i6400 and i6500.

Also test msa feature in mips-cpus.c driver test case

mgoudar committed:
Revert "Enable MSA feature by default for MIPS cpus i6400 and i6500"

This reverts commit f1b35e8ff093acaa3e41c5c9cba23b2b5346c19e.

mgoudar committed:
Revert "Revert "Enable MSA feature by default for MIPS cpus i6400 and i6500""

This reverts commit dad133765b97bc0a1de53e0817ec4af5278202c1.

keeping MSA enabled by default for cpu i6400 and i6500

mgoudar committed:
add llvm codegen test to validate MSA instructions for CPU i6500 and i6400

brad0 commented:
@mshockwave

mshockwave commented on llvm/test/CodeGen/Mips/msa/i6500-msa.ll:
```diff
@@ -0,0 +1,724 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple=mips64 -mcpu=i6500 < %s | FileCheck %s --check-prefixes=ALL
+; RUN: llc -mtriple=mips64 -mcpu=i6400 < %s | FileCheck %s --check-prefixes=ALL
```
if there is only a single CHECK prefix, please use the default one

mshockwave commented on llvm/test/CodeGen/Mips/msa/i6500-msa.ll:
I just noticed there is already `test/CodeGen/Mips/msa/arithmetic.ll` (which this file is probably based on), could add additional `RUN` lines to that file instead, in order to reuse it?

mgoudar replied:
I added RUN commands for i6500/i6400 msa test to the existing test test/CodeGen/Mips/msa/arithmetic.ll and deleted new test that was added in the prev commit.

mshockwave reviewed (commented):
Sorry I missed the update. I only have some minor comments

mgoudar committed:
Use llvm/test/CodeGen/Mips/msa/arithmetic.ll for i6500 test

* use existing llvm/test/CodeGen/Mips/msa/arithmetic.ll for
  i6500 and i6400 MSA feature test
* remove llvm/test/CodeGen/Mips/msa/i6500-msa.ll

mshockwave reviewed (approved):
LGTM, thanks

mgoudar commented:
hi @brad0 I am seeing fails in https://lab.llvm.org/buildbot/#/builders/94 post this pr merge. 
test llvm/test/CodeGen/Mips/msa/arithmetic.ll fails with a crash. I am trying to reproduce at my side

kstoimenov commented:
@mgoudar FYI I reverted this patch due to sanitizer bot failure.
